File: pipeline/verifier.py
"""The mechanical citation guarantee. Specified in docs/VERIFIER.md.

This is the module that keeps the project's central promise: no reference
reaches a reader unless the passages behind it were actually retrieved. It does
not trust the prompt, because the prompt is an instruction and the model is not
obliged to follow it.

Errors here are asymmetric. Missing a fabricated reference puts a false
citation in front of a reader. Flagging a real phrase costs one retry. The
rules lean towards detection, and the non-detection list keeps that lean from
mangling ordinary prose.

P4 ports this to Rust. TEST_VECTORS is the shared contract.
"""

import os
import re
import sqlite3
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
sys.path.insert(0, HERE)

DB = os.environ.get('TPB_INDEX_DB') or os.path.join(
    ROOT, 'src-tauri', 'resources', 'index.db')

TOKEN_RE = re.compile(r'\[P(\d+)\]')

# Book names that are also ordinary English words. These are matched
# case-sensitively so "he acts 2 ways" is not read as a citation of Acts.
AMBIGUOUS = {
    'job', 'mark', 'acts', 'numbers', 'judges', 'kings', 'song', 'songs',
    'revelation', 'chronicles', 'romans', 'hebrews', 'lamentations', 'kings',
    'proverbs', 'psalm', 'psalms', 'james', 'philemon', 'ruth', 'wisdom',
}

# A number followed by one of these is a count, not a chapter.
UNIT_NOUNS = {
    'times', 'time', 'days', 'day', 'years', 'year', 'months', 'month',
    'weeks', 'week', 'hours', 'hour', 'people', 'men', 'women', 'sons',
    'daughters', 'tribes', 'thousand', 'hundred', 'million', 'percent',
    'degrees',
}

ORDINAL_PREFIX = {
    'first': '1', '1st': '1', 'i': '1', '1': '1',
    'second': '2', '2nd': '2', 'ii': '2', '2': '2',
    'third': '3', '3rd': '3', 'iii': '3', '3': '3',
}


def _norm(tok):
    return re.sub(r'[\s.]', '', tok).lower()


# Common English names of books that neither the WEB's own long titles nor the
# TSK abbreviation table spells out. Without these, "Song of Solomon 3:1" and
# every deuterocanonical reference are invisible to Rule B: measured on
# 2026-08-26, 14 of 83 realistic book names were undetected, and 11 of the 14
# were the Deuterocanon, which is exactly what a both-canon answer cites.
ALIASES = {
    'SNG': ['Song of Songs', 'Song of Solomon', 'Canticles'],
    'ACT': ['Acts of the Apostles'],
    'WIS': ['Wisdom of Solomon', 'Wisdom'],
    'SIR': ['Ecclesiasticus', 'Sirach', 'Ben Sira'],
    'MAN': ['Prayer of Manasseh', 'Prayer of Manasses'],
    'ESG': ['Greek Esther'],
    'DAG': ['Greek Daniel'],
    '1ES': ['1 Esdras'],
    '2ES': ['2 Esdras'],
    '1MA': ['1 Maccabees'],
    '2MA': ['2 Maccabees'],
    '3MA': ['3 Maccabees'],
    '4MA': ['4 Maccabees'],
}


class BookNames(object):
    """Maps written book names onto book ids, from the index itself."""

    def __init__(self, con):
        self.by_name = {}
        # The written form each normalised key came from, so the reference
        # pattern can be rebuilt with its spaces intact. Without it a
        # multi-word name is escaped as one run of letters and can never match.
        self.written_of = {}
        self.ambiguous_ids = set()
        rows = con.execute(
            'SELECT book_id, usfm_code, name, abbrev FROM books').fetchall()
        for book_id, code, name, abbrev in rows:
            for variant in self._variants(code, name, abbrev):
                key = _norm(variant)
                if key not in self.by_name:
                    self.by_name[key] = book_id
                    self.written_of[key] = variant
                if key in AMBIGUOUS:
                    self.ambiguous_ids.add(book_id)
        # Abbreviations and common forms the source does not spell out.
        try:
            from books import TSK_ABBREV
            codes = {c: b for b, c in con.execute(
                'SELECT book_id, usfm_code FROM books')}
            for abbr, code in TSK_ABBREV.items():
                if code in codes:
                    key = _norm(abbr)
                    if key not in self.by_name:
                        self.by_name[key] = codes[code]
                        self.written_of[key] = abbr
        except Exception:  # noqa: BLE001
            pass

    @staticmethod
    def _variants(code, name, abbrev):
        out = [code, abbrev]
        # The long name is "The First Book of Moses, Commonly Called Genesis";
        # the useful part is the last word or two.
        tail = re.sub(r'^.*\bCalled\b\s*', '', name).strip()
        out.append(tail)
        out.append(name)
        # "The Song of Solomon" is how the WEB titles the book; "Song of
        # Solomon" is how anyone writes it.
        for v in (tail, name):
            if v.lower().startswith('the '):
                out.append(v[4:])
        out.extend(ALIASES.get(code, []))
        # Numeric-prefixed books: 1SA -> "1 Samuel", "First Samuel",
        # "I Samuel", all normalising to the same key.
        m = re.match(r'^([123])(.*)$', code)
        if m and tail:
            base = re.sub(r'^[123]\s*', '', tail)
            n = m.group(1)
            words = {'1': ('first', 'i'), '2': ('second', 'ii'),
                     '3': ('third', 'iii')}[n]
            for pre in (n, n + ' ') + tuple(w + ' ' for w in words):
                out.append('%s%s' % (pre, base))
        return [v for v in out if v]

    def lookup(self, written):
        key = _norm(written)
        hit = self.by_name.get(key)
        if hit is not None:
            return hit
        # "First Corinthians", "I Corinthians", "1st Corinthians" all mean
        # 1 Corinthians. Rewrite the prefix to its digit and try again. The
        # rewrite is only accepted when the result is a book we know, so
        # "Isaiah" is not mangled into "1saiah".
        for word, digit in sorted(ORDINAL_PREFIX.items(),
                                  key=lambda kv: -len(kv[0])):
            if key.startswith(word) and len(key) > len(word):
                hit = self.by_name.get(digit + key[len(word):])
                if hit is not None:
                    return hit
        return None


ORDINAL_ALT = {'1': 'First|1st|I', '2': 'Second|2nd|II', '3': 'Third|3rd|III'}


def book_alternative(written):
    """A permissive pattern for one written book name.

    The name is matched word by word so that a multi-word title keeps its
    spaces. An earlier version built the pattern from the normalised key, which
    has the spaces stripped, so "Song of Solomon 3:1" and every
    deuterocanonical reference slipped past Rule B unflagged.
    """
    parts = []
    for i, tok in enumerate(written.split()):
        m = re.match(r'^([123])(.*)$', tok) if i == 0 else None
        if m:
            pre, rest = m.group(1), m.group(2)
            head = r'(?:%s|%s)' % (pre, ORDINAL_ALT[pre])
            parts.append(head + (r'\s*\.?\s*' + re.escape(rest) if rest else ''))
        else:
            parts.append(re.escape(tok))
    # A full stop may follow any word of an abbreviated name.
    return r'\.?\s+'.join(parts)


def build_reference_re(book_names):
    """One alternation over every known written form, longest first."""
    keys = sorted(book_names.by_name, key=len, reverse=True)
    alts = []
    seen = set()
    for k in keys:
        if k in seen:
            continue
        seen.add(k)
        alts.append(book_alternative(book_names.written_of.get(k, k)))
    body = '|'.join(alts)
    return re.compile(
        r'\b(?P<book>%s)\s*\.?\s+(?P<chapter>\d{1,3})'
        r'(?:\s*[:.]\s*(?P<verse>\d{1,3})'
        r'(?:\s*[-–—]\s*(?P<verse_end>\d{1,3}))?)?' % body,
        re.IGNORECASE)


class Verifier(object):
    def __init__(self, db_path=DB):
        self.con = sqlite3.connect('file:%s?mode=ro' % db_path.replace('\\', '/'),
                                   uri=True)
        self.books = BookNames(self.con)
        self.ref_re = build_reference_re(self.books)
        self.chapter_sizes = {}

    def chapter_verses(self, book_id, chapter):
        key = (book_id, chapter)
        if key not in self.chapter_sizes:
            self.chapter_sizes[key] = {
                r[0] for r in self.con.execute(
                    'SELECT verse FROM verses WHERE book_id=? AND chapter=?',
                    (book_id, chapter))}
        return self.chapter_sizes[key]

    # -- the two rules ----------------------------------------------------

    def check(self, text, passages):
        """passages: list of dicts with 'token' and 'verse_ids'.

        Returns (verdict, violations). verdict is 'ok' or 'violation'.
        """
        sent_tokens = {p['token'] for p in passages}
        sent_verses = set()
        for p in passages:
            sent_verses.update(p['verse_ids'])

        violations = []

        # Rule A
        for m in TOKEN_RE.finditer(text):
            tok = '[P%s]' % m.group(1)
            if tok not in sent_tokens:
                violations.append({'kind': 'token', 'text': tok,
                                   'reason': 'token not in sent set',
                                   'span': m.span()})

        # Rule B
        for m in self.ref_re.finditer(text):
            written = m.group(0)
            book_written = m.group('book')
            book_id = self.books.lookup(book_written)
            if book_id is None:
                continue
            # An ambiguous book name must be capitalised to count.
            if _norm(book_written) in AMBIGUOUS and not book_written[:1].isupper():
                continue
            # A number followed by a unit noun is a count, not a chapter.
            after = text[m.end():m.end() + 24]
            nxt = re.match(r'\s+([A-Za-z]+)', after)
            if (m.group('verse') is None and nxt
                    and nxt.group(1).lower() in UNIT_NOUNS):
                continue

            chapter = int(m.group('chapter'))
            if m.group('verse') is None:
                wanted = {book_id * 1000000 + chapter * 1000 + v
                          for v in self.chapter_verses(book_id, chapter)}
                reason = 'whole chapter not in sent set'
                if not wanted:
                    reason = 'chapter not in this text'
            else:
                v1 = int(m.group('verse'))
                v2 = int(m.group('verse_end') or v1)
                wanted = {book_id * 1000000 + chapter * 1000 + v
                          for v in range(v1, v2 + 1)}
                reason = 'verses not in sent set'
            if not wanted or not wanted <= sent_verses:
                violations.append({'kind': 'reference', 'text': written,
                                   'reason': reason, 'span': m.span()})

        return ('violation' if violations else 'ok'), violations

    @staticmethod
    def strip(text, violations):
        """Remove offending spans, right to left so offsets stay valid."""
        out = text
        for v in sorted(violations, key=lambda v: -v['span'][0]):
            a, b = v['span']
            out = out[:a] + out[b:]
        return re.sub(r'[ \t]{2,}', ' ', out)

    @staticmethod
    def failure_note(violations):
        """The text handed back to the model on the retry."""
        parts = []
        for v in violations:
            parts.append('%s "%s" (%s)' % (v['kind'], v['text'], v['reason']))
        return '; '.join(parts)

    def fallback(self, passages):
        """PLAN 5.6's fallback: passages grouped by book, with the note."""
        by_book = {}
        for p in passages:
            bid = p['verse_ids'][0] // 1000000
            by_book.setdefault(bid, []).append(p)
        names = dict(self.con.execute('SELECT book_id, abbrev FROM books'))
        order = sorted(by_book)
        lines = ['A synthesis could not be produced for this question. '
                 'These are the passages that were found.']
        for bid in order:
            lines.append('')
            lines.append(names.get(bid, '?'))
            for p in sorted(by_book[bid], key=lambda p: p['verse_ids'][0]):
                lines.append('  %s' % p.get('ref', p['token']))
        return '\n'.join(lines)


# ---------------------------------------------------------------------------
# Test vectors. The contract shared with the Rust port. Sent set:
#   [P1] Matthew 6:25-34   [P2] Philippians 4:6-7
#   [P3] Psalm 23:1-4      [P4] 1 Peter 5:6-7
# ---------------------------------------------------------------------------

SENT_SPEC = [
    ('P1', 'MAT', 6, 25, 34),
    ('P2', 'PHP', 4, 6, 7),
    ('P3', 'PSA', 23, 1, 4),
    ('P4', '1PE', 5, 6, 7),
]

TEST_VECTORS = [
    # 1-8 must pass
    ('Jesus tells us not to worry [P1].', 'ok'),
    ('Anxiety is answered by prayer [P1] [P2].', 'ok'),
    ('As Matthew 6:25 says, do not be anxious [P1].', 'ok'),
    ('See Mat 6:26 for the birds [P1].', 'ok'),
    ('The passage Matthew 6:25-34 covers this [P1].', 'ok'),
    ('The passage Matthew 6:25–34 covers this [P1].', 'ok'),
    ('The text speaks plainly about worry and trust.', 'ok'),
    ('Trust God ([P4]), and be humble.', 'ok'),
    # 9-17 must be flagged
    ('Consider also [P9], which speaks to this.', 'violation'),
    ('See [P0] for more.', 'violation'),
    ('As John 3:16 tells us, God loved the world.', 'violation'),
    ('Matthew 5:3 speaks of the poor in spirit.', 'violation'),
    ('Read Matthew 6:25-40 for the whole argument.', 'violation'),
    ('Psalm 23 is the shepherd psalm.', 'violation'),
    ('1 Corinthians 13:4 defines love.', 'violation'),
    ('First Corinthians 13 defines love.', 'violation'),
    ('I Corinthians 13 defines love.', 'violation'),
    # 18-25 must not be flagged
    ('He rose on the third day, as the passages say [P1].', 'ok'),
    ('Forgive seven times, and then seventy [P1].', 'ok'),
    ('He could quote chapter and verse on it [P2].', 'ok'),
    ('Job suffered greatly, yet he trusted [P3].', 'ok'),
    ('Mark was a companion on the journey [P1].', 'ok'),
    ('He acts 2 ways depending on who is watching [P2].', 'ok'),
    ('Judges 12 times refused to listen to the warning [P1].', 'ok'),
    ('Numbers and Kings and Acts all tell parts of the story [P3].', 'ok'),
    # 26-31 multi-word and deuterocanonical names must be flagged. Before
    # 2026-08-26 every one of these passed as clean prose.
    ('Song of Solomon 2:1 speaks of love.', 'violation'),
    ('Song of Songs 2:1 speaks of love.', 'violation'),
    ('The Acts of the Apostles 2:38 records the sermon.', 'violation'),
    ('Wisdom of Solomon 3:1 says the souls of the righteous are in God hands.',
     'violation'),
    ('1 Maccabees 2:15 tells of the revolt.', 'violation'),
    ('Sirach 3:1 counsels children to honour their parents.', 'violation'),
    # 32-35 the same names in ordinary prose, with no chapter number, must not
    # be flagged. Detection must not cost the reader a retry over a sentence.
    ('The song of songs was sung by the whole congregation [P3].', 'ok'),
    ('The acts of the apostles were many and are still told [P1].', 'ok'),
    ('Wisdom builds her house and calls out to the simple [P2].', 'ok'),
    ('He read the prayer of Manasseh aloud at the vigil [P4].', 'ok'),
]


def sent_set(con):
    """Build the test sent set from the real index."""
    out = []
    for token, code, ch, v1, v2 in SENT_SPEC:
        bid = con.execute('SELECT book_id FROM books WHERE usfm_code=?',
                          (code,)).fetchone()[0]
        ids = [bid * 1000000 + ch * 1000 + v for v in range(v1, v2 + 1)]
        ids = [i for i in ids if con.execute(
            'SELECT 1 FROM verses WHERE verse_id=?', (i,)).fetchone()]
        out.append({'token': '[%s]' % token, 'verse_ids': ids,
                    'ref': '%s %d:%d-%d' % (code, ch, v1, v2)})
    return out


def run_vectors(db_path=DB):
    v = Verifier(db_path)
    passages = sent_set(v.con)
    results = []
    for i, (text, expected) in enumerate(TEST_VECTORS, start=1):
        got, violations = v.check(text, passages)
        results.append((i, text, expected, got, violations))
    return results


if __name__ == '__main__':
    sys.stdout.reconfigure(encoding='utf-8')
    bad = 0
    for i, text, expected, got, violations in run_vectors():
        mark = 'ok  ' if got == expected else 'FAIL'
        if got != expected:
            bad += 1
        print('%s %2d  expected %-9s got %-9s  %s' % (mark, i, expected, got,
                                                      text[:58]))
        if got != expected and violations:
            for vi in violations:
                print('         -> %s %r (%s)' % (vi['kind'], vi['text'],
                                                  vi['reason']))
    print()
    print('%d/%d vectors pass' % (len(TEST_VECTORS) - bad, len(TEST_VECTORS)))

File: src-tauri/core/src/verifier.rs
//! The mechanical citation guarantee, specified in docs/VERIFIER.md.
//!
//! A port of pipeline/verifier.py. The 35 test vectors in that file are the
//! shared contract and this implementation runs them too; core/tests/ holds
//! them as a fixture so neither side can quietly diverge from the other.
//!
//! Errors here are asymmetric. Missing a fabricated reference puts a false
//! citation in front of a reader. Flagging a real phrase costs one retry. The
//! rules lean towards detection, and the non-detection list keeps that lean
//! from mangling ordinary prose.

use std::collections::{HashMap, HashSet};

use regex::Regex;

use crate::index::Index;
use crate::tsk_abbrev::TSK_ABBREV;

/// Book names that are also ordinary English words. Matched case-sensitively,
/// so "he acts 2 ways" is not read as a citation of Acts.
pub const AMBIGUOUS: &[&str] = &[
    "job",
    "mark",
    "acts",
    "numbers",
    "judges",
    "kings",
    "song",
    "songs",
    "revelation",
    "chronicles",
    "romans",
    "hebrews",
    "lamentations",
    "proverbs",
    "psalm",
    "psalms",
    "james",
    "philemon",
    "ruth",
    "wisdom",
];

/// A number followed by one of these is a count, not a chapter.
pub const UNIT_NOUNS: &[&str] = &[
    "times", "time", "days", "day", "years", "year", "months", "month", "weeks", "week", "hours",
    "hour", "people", "men", "women", "sons", "daughters", "tribes", "thousand", "hundred",
    "million", "percent", "degrees",
];

const ORDINAL_PREFIX: &[(&str, &str)] = &[
    ("first", "1"),
    ("1st", "1"),
    ("i", "1"),
    ("1", "1"),
    ("second", "2"),
    ("2nd", "2"),
    ("ii", "2"),
    ("2", "2"),
    ("third", "3"),
    ("3rd", "3"),
    ("iii", "3"),
    ("3", "3"),
];

/// Common English names of books that neither the WEB's own long titles nor the
/// TSK abbreviation table spells out. Without these, "Song of Solomon 3:1" and
/// every deuterocanonical reference are invisible to Rule B.
const ALIASES: &[(&str, &[&str])] = &[
    ("SNG", &["Song of Songs", "Song of Solomon", "Canticles"]),
    ("ACT", &["Acts of the Apostles"]),
    ("WIS", &["Wisdom of Solomon", "Wisdom"]),
    ("SIR", &["Ecclesiasticus", "Sirach", "Ben Sira"]),
    ("MAN", &["Prayer of Manasseh", "Prayer of Manasses"]),
    ("ESG", &["Greek Esther"]),
    ("DAG", &["Greek Daniel"]),
    ("1ES", &["1 Esdras"]),
    ("2ES", &["2 Esdras"]),
    ("1MA", &["1 Maccabees"]),
    ("2MA", &["2 Maccabees"]),
    ("3MA", &["3 Maccabees"]),
    ("4MA", &["4 Maccabees"]),
];

pub fn norm(tok: &str) -> String {
    tok.chars().filter(|c| !c.is_whitespace() && *c != '.').flat_map(|c| c.to_lowercase()).collect()
}

fn is_ambiguous(key: &str) -> bool {
    AMBIGUOUS.contains(&key)
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Kind {
    Token,
    Reference,
}

impl Kind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Kind::Token => "token",
            Kind::Reference => "reference",
        }
    }
}

#[derive(Clone, Debug)]
pub struct Violation {
    pub kind: Kind,
    pub text: String,
    pub reason: String,
    /// Byte offsets, which is what `strip` needs.
    pub span: (usize, usize),
    /// Character offsets, which is what Python's re reports. The two differ
    /// the moment an answer contains an en dash or a curly quote, and the
    /// parity check against the Python records compares these.
    pub char_span: (usize, usize),
}

fn char_span(text: &str, span: (usize, usize)) -> (usize, usize) {
    let start = text[..span.0].chars().count();
    (start, start + text[span.0..span.1].chars().count())
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Verdict {
    Ok,
    Violation,
}

impl Verdict {
    pub fn as_str(&self) -> &'static str {
        match self {
            Verdict::Ok => "ok",
            Verdict::Violation => "violation",
        }
    }
}

/// One passage as it was sent to the model.
#[derive(Clone, Debug)]
pub struct Sent {
    pub token: String,
    pub reference: String,
    pub verse_ids: Vec<i64>,
}

struct BookNames {
    by_name: HashMap<String, i64>,
    /// The written form each normalised key came from, so the reference pattern
    /// can be rebuilt with its spaces intact.
    written_of: HashMap<String, String>,
}

impl BookNames {
    fn new(index: &Index) -> Self {
        let mut by_name: HashMap<String, i64> = HashMap::new();
        let mut written_of: HashMap<String, String> = HashMap::new();
        let mut insert = |key: String, written: String, book_id: i64| {
            if !by_name.contains_key(&key) {
                by_name.insert(key.clone(), book_id);
                written_of.insert(key, written);
            }
        };
        for b in &index.books {
            for variant in variants(&b.usfm_code, &b.name, &b.abbrev) {
                insert(norm(&variant), variant, b.book_id);
            }
        }
        let by_code: HashMap<&str, i64> =
            index.books.iter().map(|b| (b.usfm_code.as_str(), b.book_id)).collect();
        for (abbr, code) in TSK_ABBREV {
            if let Some(&bid) = by_code.get(code) {
                insert(norm(abbr), abbr.to_string(), bid);
            }
        }
        BookNames { by_name, written_of }
    }

    fn lookup(&self, written: &str) -> Option<i64> {
        let key = norm(written);
        if let Some(&b) = self.by_name.get(&key) {
            return Some(b);
        }
        // "First Corinthians", "I Corinthians", "1st Corinthians" all mean
        // 1 Corinthians. Rewrite the prefix to its digit and try again. The
        // rewrite is accepted only when the result names a real book, so
        // "Isaiah" is not mangled into "1saiah".
        let mut prefixes: Vec<&(&str, &str)> = ORDINAL_PREFIX.iter().collect();
        prefixes.sort_by_key(|(w, _)| std::cmp::Reverse(w.len()));
        for (word, digit) in prefixes {
            if key.starts_with(word) && key.len() > word.len() {
                let rewritten = format!("{}{}", digit, &key[word.len()..]);
                if let Some(&b) = self.by_name.get(&rewritten) {
                    return Some(b);
                }
            }
        }
        None
    }
}

fn variants(code: &str, name: &str, abbrev: &str) -> Vec<String> {
    let mut out = vec![code.to_string(), abbrev.to_string()];
    // The long name is "The First Book of Moses, Commonly Called Genesis"; the
    // useful part is the last word or two.
    let tail = match name.find("Called ") {
        Some(i) => name[i + "Called ".len()..].trim().to_string(),
        None => name.trim().to_string(),
    };
    out.push(tail.clone());
    out.push(name.to_string());
    // "The Song of Solomon" is how the WEB titles the book; "Song of Solomon"
    // is how anyone writes it.
    for v in [&tail, &name.to_string()] {
        if v.len() > 4 && v[..4].eq_ignore_ascii_case("the ") {
            out.push(v[4..].to_string());
        }
    }
    for (c, names) in ALIASES {
        if *c == code {
            out.extend(names.iter().map(|s| s.to_string()));
        }
    }
    // Numeric-prefixed books: 1SA -> "1Samuel", "1 Samuel", "first Samuel",
    // "i Samuel", all normalising to the same key.
    let first = code.chars().next().unwrap_or(' ');
    if ('1'..='3').contains(&first) && !tail.is_empty() {
        let base = tail.trim_start_matches(['1', '2', '3']).trim_start().to_string();
        let words: [&str; 2] = match first {
            '1' => ["first", "i"],
            '2' => ["second", "ii"],
            _ => ["third", "iii"],
        };
        let mut prefixes = vec![first.to_string(), format!("{} ", first)];
        prefixes.extend(words.iter().map(|w| format!("{} ", w)));
        for p in prefixes {
            out.push(format!("{}{}", p, base));
        }
    }
    out.into_iter().filter(|v| !v.is_empty()).collect()
}

const ORDINAL_ALT: &[(&str, &str)] =
    &[("1", "First|1st|I"), ("2", "Second|2nd|II"), ("3", "Third|3rd|III")];

/// A permissive pattern for one written book name.
///
/// The name is matched word by word so that a multi-word title keeps its
/// spaces. Building the pattern from the normalised key instead, which has the
/// spaces stripped, is what let "Song of Solomon 3:1" and every
/// deuterocanonical reference slip past Rule B before 2026-08-26.
fn book_alternative(written: &str) -> String {
    let mut parts: Vec<String> = Vec::new();
    for (i, tok) in written.split_whitespace().enumerate() {
        let first = tok.chars().next().unwrap_or(' ');
        if i == 0 && ('1'..='3').contains(&first) {
            let rest = &tok[first.len_utf8()..];
            let alt = ORDINAL_ALT.iter().find(|(d, _)| d.chars().next() == Some(first)).unwrap().1;
            let head = format!("(?:{}|{})", first, alt);
            if rest.is_empty() {
                parts.push(head);
            } else {
                parts.push(format!("{}\\s*\\.?\\s*{}", head, regex::escape(rest)));
            }
        } else {
            parts.push(regex::escape(tok));
        }
    }
    // A full stop may follow any word of an abbreviated name.
    parts.join("\\.?\\s+")
}

fn build_reference_re(books: &BookNames) -> Regex {
    let mut keys: Vec<&String> = books.by_name.keys().collect();
    // Longest first, and by the key itself for equal lengths so the alternation
    // is built in one order on every run.
    keys.sort_by(|a, b| b.len().cmp(&a.len()).then(a.cmp(b)));
    let alts: Vec<String> = keys
        .into_iter()
        .map(|k| book_alternative(books.written_of.get(k).map(|s| s.as_str()).unwrap_or(k)))
        .collect();
    let body = alts.join("|");
    let pattern = format!(
        r"(?i)\b(?P<book>{})\s*\.?\s+(?P<chapter>\d{{1,3}})(?:\s*[:.]\s*(?P<verse>\d{{1,3}})(?:\s*[-\u{{2013}}\u{{2014}}]\s*(?P<verse_end>\d{{1,3}}))?)?",
        body
    );
    regex::RegexBuilder::new(&pattern)
        .size_limit(64 * 1024 * 1024)
        .dfa_size_limit(64 * 1024 * 1024)
        .build()
        .expect("reference pattern")
}

pub struct Verifier {
    books: BookNames,
    ref_re: Regex,
    token_re: Regex,
    after_re: Regex,
}

impl Verifier {
    pub fn new(index: &Index) -> Self {
        let books = BookNames::new(index);
        let ref_re = build_reference_re(&books);
        Verifier {
            books,
            ref_re,
            token_re: Regex::new(r"\[P(\d+)\]").unwrap(),
            after_re: Regex::new(r"^\s+([A-Za-z]+)").unwrap(),
        }
    }

    /// Rule A and Rule B over one model output.
    pub fn check(&self, index: &Index, text: &str, passages: &[Sent]) -> (Verdict, Vec<Violation>) {
        let sent_tokens: HashSet<&str> = passages.iter().map(|p| p.token.as_str()).collect();
        let mut sent_verses: HashSet<i64> = HashSet::new();
        for p in passages {
            sent_verses.extend(p.verse_ids.iter().copied());
        }

        let mut violations: Vec<Violation> = Vec::new();

        // Rule A: every token must be one that was sent.
        for m in self.token_re.captures_iter(text) {
            let whole = m.get(0).unwrap();
            let tok = format!("[P{}]", &m[1]);
            if !sent_tokens.contains(tok.as_str()) {
                violations.push(Violation {
                    kind: Kind::Token,
                    text: tok,
                    reason: "token not in sent set".to_string(),
                    span: (whole.start(), whole.end()),
                    char_span: char_span(text, (whole.start(), whole.end())),
                });
            }
        }

        // Rule B: every free-text reference must resolve inside the sent set.
        for caps in self.ref_re.captures_iter(text) {
            let whole = caps.get(0).unwrap();
            let written = whole.as_str();
            let book_written = caps.name("book").unwrap().as_str();
            let book_id = match self.books.lookup(book_written) {
                Some(b) => b,
                None => continue,
            };
            // An ambiguous book name must be capitalised to count.
            if is_ambiguous(&norm(book_written))
                && !book_written.chars().next().map(|c| c.is_uppercase()).unwrap_or(false)
            {
                continue;
            }
            // A number followed by a unit noun is a count, not a chapter.
            // Python looks at the next 24 characters; a byte slice would cut
            // a multi-byte character in half.
            let after: String = text[whole.end()..].chars().take(24).collect();
            if caps.name("verse").is_none() {
                if let Some(nxt) = self.after_re.captures(&after) {
                    if UNIT_NOUNS.contains(&nxt[1].to_lowercase().as_str()) {
                        continue;
                    }
                }
            }

            let chapter: i64 = caps["chapter"].parse().unwrap_or(0);
            let (wanted, reason): (HashSet<i64>, &str) = match caps.name("verse") {
                None => {
                    let verses = index.chapter_verses(book_id, chapter);
                    let set: HashSet<i64> = verses
                        .iter()
                        .map(|v| book_id * 1_000_000 + chapter * 1000 + v)
                        .collect();
                    let reason = if set.is_empty() {
                        "chapter not in this text"
                    } else {
                        "whole chapter not in sent set"
                    };
                    (set, reason)
                }
                Some(v1) => {
                    let v1: i64 = v1.as_str().parse().unwrap_or(0);
                    let v2: i64 =
                        caps.name("verse_end").map(|m| m.as_str().parse().unwrap_or(v1)).unwrap_or(v1);
                    let set: HashSet<i64> =
                        (v1..=v2).map(|v| book_id * 1_000_000 + chapter * 1000 + v).collect();
                    (set, "verses not in sent set")
                }
            };
            if wanted.is_empty() || !wanted.is_subset(&sent_verses) {
                violations.push(Violation {
                    kind: Kind::Reference,
                    text: written.to_string(),
                    reason: reason.to_string(),
                    span: (whole.start(), whole.end()),
                    char_span: char_span(text, (whole.start(), whole.end())),
                });
            }
        }

        let verdict = if violations.is_empty() { Verdict::Ok } else { Verdict::Violation };
        (verdict, violations)
    }

    /// Remove offending spans, right to left so offsets stay valid.
    ///
    /// A stripped output is never shown to a reader. It exists so the retry
    /// prompt can quote what was wrong.
    pub fn strip(text: &str, violations: &[Violation]) -> String {
        let mut spans: Vec<(usize, usize)> = violations.iter().map(|v| v.span).collect();
        spans.sort_by(|a, b| b.0.cmp(&a.0));
        let mut out = text.to_string();
        for (a, b) in spans {
            if b <= out.len() {
                out.replace_range(a..b, "");
            }
        }
        // Collapse the runs of spaces that removal leaves behind.
        let squeeze = Regex::new(r"[ \t]{2,}").unwrap();
        squeeze.replace_all(&out, " ").into_owned()
    }

    /// The text handed back to the model on the retry.
    pub fn failure_note(violations: &[Violation]) -> String {
        violations
            .iter()
            .map(|v| format!("{} \"{}\" ({})", v.kind.as_str(), v.text, v.reason))
            .collect::<Vec<_>>()
            .join("; ")
    }

    /// PLAN 5.6's fallback: the retrieved passages grouped by book, with the
    /// one-line note. A normal outcome, not an error state.
    pub fn fallback(index: &Index, passages: &[Sent]) -> String {
        let mut by_book: HashMap<i64, Vec<&Sent>> = HashMap::new();
        for p in passages {
            if let Some(&first) = p.verse_ids.first() {
                by_book.entry(first / 1_000_000).or_default().push(p);
            }
        }
        let mut order: Vec<i64> = by_book.keys().copied().collect();
        order.sort();
        let mut lines = vec![
            "A synthesis could not be produced for this question. These are the \
             passages that were found."
                .to_string(),
        ];
        for bid in order {
            lines.push(String::new());
            lines.push(index.abbrev(bid).to_string());
            let mut ps = by_book.remove(&bid).unwrap();
            ps.sort_by_key(|p| p.verse_ids.first().copied().unwrap_or(0));
            for p in ps {
                let label = if p.reference.is_empty() { &p.token } else { &p.reference };
                lines.push(format!("  {}", label));
            }
        }
        lines.join("\n")
    }
}

/// The tokens an answer cites, sorted and deduplicated.
pub fn cited_tokens(text: &str) -> Vec<String> {
    let re = Regex::new(r"\[P\d+\]").unwrap();
    let mut out: Vec<String> = re.find_iter(text).map(|m| m.as_str().to_string()).collect();
    out.sort();
    out.dedup();
    out
}
